source/robot_model/src/Model.cpp

```diff
@@ -378,39 +370,99 @@ Model::inverse_kinematics(const state_representation::CartesianPose& cartesian_p
-                          const std::string& frame_name,
-                          const InverseKinematicsParameters& parameters) {
+                          const InverseKinematicsParameters& parameters,
+                          const std::string& frame_name) {
   Eigen::VectorXd q(pinocchio::neutral(this->robot_model_));
   state_representation::JointPositions positions(this->get_robot_name(), this->get_joint_frames(), q);
-  return this->inverse_kinematics(cartesian_pose, positions, frame_name, parameters);
+  return this->inverse_kinematics(cartesian_pose, positions, parameters, frame_name);
 }
 
 state_representation::CartesianTwist Model::forward_velocity(const state_representation::JointState& joint_state) {
   return this->compute_jacobian(joint_state) * static_cast<state_representation::JointVelocities>(joint_state);
 }
 
+void Model::check_inverse_velocity_arguments(const std::vector<state_representation::CartesianTwist>& cartesian_twists,
+                                             const state_representation::JointPositions& joint_positions,
+                                             const std::vector<std::string>& frame_names) {
+  if (cartesian_twists.size() != frame_names.size()) {
+    throw (std::invalid_argument("The number of provided twists and frame names does not match"));
+  }
+  if (joint_positions.get_size() != this->get_number_of_joints()) {
+    throw (exceptions::InvalidJointStateSizeException(joint_positions.get_size(), this->get_number_of_joints()));
+  }
+  for (auto& frame_name : frame_names) {
+    if (!this->robot_model_.existFrame(frame_name)) {
+      throw (exceptions::FrameNotFoundException(frame_name));
+    }
+  }
+}
+
 state_representation::JointVelocities
 Model::inverse_velocity(const std::vector<state_representation::CartesianTwist>& cartesian_twists,
                         const state_representation::JointPositions& joint_positions,
-                        const InverseVelocityParameters& parameters) {
-  this->init_qp_solver(parameters);
+                        const std::vector<std::string>& frame_names) {
+  // sanity check
+  this->check_inverse_velocity_arguments(cartesian_twists, joint_positions, frame_names);
+
+  const unsigned int nb_joints = this->get_number_of_joints();
+  // the velocity vector contains position of the intermediate frame and full pose of the end-effector
+  Eigen::VectorXd dX(3 * cartesian_twists.size() + 3);
+  Eigen::MatrixXd jacobian(3 * cartesian_twists.size() + 3, nb_joints);
+  for (unsigned int i = 0; i < cartesian_twists.size() - 1; ++i) {
+    state_representation::CartesianTwist twist = cartesian_twists[i];
+    // extract only the linear velocity for intermediate points
+    dX.segment<3>(3 * i) = twist.get_linear_velocity();
+    jacobian.block(3 * i, 0, 3 * i + 3, nb_joints) =
+        this->compute_jacobian(joint_positions, frame_names.at(i)).data().block(0, 0, 3, nb_joints);
+  }
+  // full twist for the end-effector
```

Comment:
This is a little detail but it doesn't necessarily have to be the end-effector right?

Reply:
indeed it is just the last one of the frames

Reply:
corrected